File: internal/usecases/dto/refer.go
package dto

// =====================
// ReferForm は参照リンクのフォーム DTO である。
type ReferForm struct {
	ID        int64        `json:"id"`
	Title     string       `json:"title"`
	URL       string       `json:"url"`
	CreatedAt *string      `json:"createdAt,omitempty"`
	UpdatedAt *string      `json:"updatedAt,omitempty"`
	DeletedAt *string      `json:"deletedAt,omitempty"`
	Answers   []AnswerForm `json:"answers,omitempty"`
}

File: static/js/models/Refer.js
/**
 * Refer model class.
 */

import { Answer } from '/static/js/models/Answer.js';
import { formToApiJson } from '/static/js/models/helpers.js';

export class Refer {
  /**
   * @param {number}   id
   * @param {string}   title
   * @param {string}   url
   * @param {Answer[]} [answers]
   * @param {string|null} createdAt
   * @param {string|null} updatedAt
   */
  constructor(id, title, url, answers = [], createdAt = null, updatedAt = null) {
    this.id = id;
    this.title = title;
    this.url = url;
    this.answers = answers;
    this.createdAt = createdAt ? new Date(createdAt) : null;
    this.updatedAt = updatedAt ? new Date(updatedAt) : null;
  }

  /** @param {Object} json @returns {Refer} */
  static fromJSON(json) {
    return new Refer(
      json.id,
      json.title,
      json.url,
      (json.answers ?? []).map(Answer.fromJSON),
      json.createdAt ?? null,
      json.updatedAt ?? null,
    );
  }

  /** @returns {Object} */
  static toModel(form) {
    return formToApiJson(form);
  }
}
